File: discovery/models/domain.py
"""Domain and DNS related models"""
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field, HttpUrl
from datetime import datetime


class PortScanResult(BaseModel):
    """Port scan result for a single port"""
    port: int = Field(..., description="Port number (1-65535)")
    protocol: str = Field(default="tcp", description="Protocol (tcp/udp)")
    state: str = Field(..., description="Port state (open/closed/filtered)")
    service: Optional[str] = Field(default=None, description="Service name (http, ssh, etc.)")
    version: Optional[str] = Field(default=None, description="Service version if detected")
    discovered_at: datetime = Field(default_factory=datetime.utcnow)


class DNSRecords(BaseModel):
    """DNS records for a domain"""
    a: List[str] = Field(default_factory=list, description="A records (IPv4)")
    aaaa: List[str] = Field(default_factory=list, description="AAAA records (IPv6)")
    mx: List[str] = Field(default_factory=list, description="MX records")
    txt: List[str] = Field(default_factory=list, description="TXT records")
    ns: List[str] = Field(default_factory=list, description="NS records")
    cname: Optional[str] = Field(default=None, description="CNAME record")


class Subdomain(BaseModel):
    """Discovered subdomain information"""
    name: str = Field(..., description="Subdomain name")
    ips: List[str] = Field(default_factory=list, description="Resolved IP addresses")
    status: str = Field(default="unknown", description="Status: live|dead|unknown")
    dns_records: Optional[DNSRecords] = Field(default=None, description="DNS records")
    cloud_provider: Optional[str] = Field(default=None, description="Cloud provider if detected")
    cdn: Optional[str] = Field(default=None, description="CDN provider if detected")
    discovered_via: str = Field(..., description="Discovery source")
    discovered_at: datetime = Field(default_factory=datetime.utcnow)

    # Port scanning results
    open_ports: List[PortScanResult] = Field(default_factory=list, description="Open ports discovered")
    total_ports_scanned: int = Field(default=0, description="Total ports scanned")
    open_ports_count: int = Field(default=0, description="Count of open ports")


class WHOISData(BaseModel):
    """WHOIS registration data"""
    registrar: Optional[str] = None
    creation_date: Optional[datetime] = None
    expiration_date: Optional[datetime] = None
    name_servers: List[str] = Field(default_factory=list)
    status: List[str] = Field(default_factory=list)
    emails: List[str] = Field(default_factory=list)
    raw_data: Optional[str] = Field(default=None, description="Raw WHOIS response")


class DomainInfo(BaseModel):
    """Complete domain intelligence"""
    root_domain: str = Field(..., description="Root domain being analyzed")
    subdomains: List[Subdomain] = Field(default_factory=list)
    dns_records: Optional[DNSRecords] = Field(default=None, description="Root domain DNS")
    whois: Optional[WHOISData] = None
    total_subdomains: int = Field(default=0, description="Total discovered subdomains")
    live_subdomains: int = Field(default=0, description="Live/responsive subdomains")

File: node-implementation/src/models/domain.ts
/**
 * Domain-related data models and schemas
 */

import { z } from 'zod';

/**
 * Port scan result for a single port
 */
export const portScanResultSchema = z.object({
  /** Port number (1-65535) */
  port: z.number().int().min(1).max(65535),
  /** Protocol (tcp/udp) */
  protocol: z.enum(['tcp', 'udp']).default('tcp'),
  /** Port state (open/closed/filtered) */
  state: z.enum(['open', 'closed', 'filtered']),
  /** Service name (http, ssh, etc.) */
  service: z.string().optional(),
  /** Service version if detected */
  version: z.string().optional(),
  /** Timestamp when port was discovered */
  discoveredAt: z.date().default(() => new Date()),
});

export type PortScanResult = z.infer<typeof portScanResultSchema>;

/**
 * DNS records for a domain
 */
export const dnsRecordsSchema = z.object({
  /** A records (IPv4) */
  a: z.array(z.string()).default([]),
  /** AAAA records (IPv6) */
  aaaa: z.array(z.string()).default([]),
  /** MX records */
  mx: z.array(z.string()).default([]),
  /** TXT records */
  txt: z.array(z.string()).default([]),
  /** NS records */
  ns: z.array(z.string()).default([]),
  /** CNAME record */
  cname: z.string().optional(),
});

export type DNSRecords = z.infer<typeof dnsRecordsSchema>;

/**
 * WHOIS registration data
 */
export const whoisDataSchema = z.object({
  /** Domain registrar */
  registrar: z.string().optional(),
  /** Domain creation date */
  creationDate: z.date().optional(),
  /** Domain expiration date */
  expirationDate: z.date().optional(),
  /** Name servers */
  nameServers: z.array(z.string()).default([]),
  /** Domain status codes */
  status: z.array(z.string()).default([]),
  /** Contact emails */
  emails: z.array(z.string()).default([]),
  /** Raw WHOIS response */
  rawData: z.string().optional(),
});

export type WHOISData = z.infer<typeof whoisDataSchema>;

/**
 * Discovered subdomain information
 */
export const subdomainSchema = z.object({
  /** Subdomain name */
  name: z.string(),
  /** Resolved IP addresses */
  ips: z.array(z.string()).default([]),
  /** Status: live|dead|unknown */
  status: z.enum(['live', 'dead', 'unknown']).default('unknown'),
  /** DNS records */
  dnsRecords: dnsRecordsSchema.optional(),
  /** Cloud provider if detected */
  cloudProvider: z.string().optional(),
  /** CDN provider if detected */
  cdnProvider: z.string().optional(),
  /** Discovery source (subfinder, etc.) */
  discoveredVia: z.string(),
  /** Timestamp when subdomain was discovered */
  discoveredAt: z.date().default(() => new Date()),
  /** Open ports discovered */
  openPorts: z.array(portScanResultSchema).default([]),
  /** Total ports scanned */
  totalPortsScanned: z.number().int().default(0),
  /** Count of open ports */
  openPortsCount: z.number().int().default(0),
  /** Associated services */
  services: z.array(z.any()).optional(),
});

export type Subdomain = z.infer<typeof subdomainSchema>;

/**
 * Complete domain intelligence
 */
export const domainInfoSchema = z.object({
  /** Root domain being analyzed */
  rootDomain: z.string(),
  /** Discovered subdomains */
  subdomains: z.array(subdomainSchema).default([]),
  /** Root domain DNS records */
  dnsRecords: dnsRecordsSchema.optional(),
  /** WHOIS data */
  whois: whoisDataSchema.optional(),
  /** Total discovered subdomains */
  totalSubdomains: z.number().int().default(0),
  /** Live/responsive subdomains */
  liveSubdomains: z.number().int().default(0),
});

export type DomainInfo = z.infer<typeof domainInfoSchema>;
